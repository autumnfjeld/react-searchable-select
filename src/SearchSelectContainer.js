import React, { Component } from 'react';
import axios from 'axios';
import CharacterInput from './CharacterInput';
import SelectableResultList from './SelectableResultList';


/**
 * Class representing a component with input form and list
 * Holds state for children that share data
 * @extends React.Component
 */
class SearchSelectContainer extends Component {
    constructor(props) {
        super(props);
        this.state = { 
            characterName: '',
            characterList: []
        };

   		this.handleCharacterNameChange = this.handleCharacterNameChange.bind(this);        
    }

    /**
     * A React lifecycle method that fires after the component is updated. Required to capture the correct state 
     * in the component lifecycle after component updates.  Good place for network requests by comparing current 
     * props/state to previous props/state
     * @param  {Object} prevProps - placeholder, not used
     * @param  {Object} prevState - the previous component state before component update
     */
    componentDidUpdate(prevProps, prevState) {
        // TODO: if network connection is slow, handle timing....waste to fetch initial typing, wait for pause or ?
        if (prevState.characterName !== this.state.characterName) {
            getCharacterList();
        }
    }	

    /** 
     * Callback for CharacterInput <input>'s on'
     */
	handleCharacterNameChange(characterName) {
	   	this.setState({'characterName': characterName});
	}  

	/**
	 * Calls the proxy server set up in api/index.js and sets state.characterList
	 */
    getCharacterList() {
    	// TODO: figure out better way to handle this = self
        var self = this;
        const url = 'http://localhost:1111/characters?nameStartsWith=' + this.state.characterName;
        // TODO: look into axios.CancelToken
        this.serverRequest = axios.get(url)
                .then(res => {
                    const characterList = res.data.results;
                    self.setState({characterList});
                });        
    } 	

    render() {
		const characterName = this.state.characterName;
		const characterList = this.state.characterList;
        return (
        	<div className='search-select-container'>
	            <CharacterInput 
	                characterName={characterName} 
	                handleChange={this.handleCharacterNameChange}
	            />
	            <SelectableResultList 
	            	characterList={characterList}
	            />
        	</div>
        )
    };	

}

export default SearchSelectContainer;