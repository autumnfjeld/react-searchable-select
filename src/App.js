import React, { Component } from 'react';
import SearchSelectContainer from './SearchSelectContainer';
import './app.css';

class App extends Component {
    render() {
      return (
            <div className='app-container'>
                <header className='app-header'>
                    <h1 className='app-title'>MARVEL</h1>
                    <div>
                        <SearchSelectContainer />
                    </div>
                </header> 
                <div className='app-content'>
                </div>
            </div>
    );
  }
}

export default App;
