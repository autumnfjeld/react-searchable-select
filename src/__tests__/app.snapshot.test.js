'use strict';

import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from '../App';
import SearchSelectContainer from '../SearchSelectContainer';

/**
 * Purpose of snapshot testing: http://facebook.github.io/jest/docs/en/snapshot-testing.html
 * A typical snapshot test case for a mobile app renders a UI component, takes a screenshot, then compares it to a 
 * reference image stored alongside the test. The test will fail if the two images do not match: either the change is 
 * unexpected, or the screenshot needs to be updated to the new version of the UI component.
 *
 * Snapshot tests preserve a starting point, but don’t 
 * give any indication of developer intent to future developers. Like any other test, if the initial snapshot passes 
 * with bugs, or doesn’t capture the full range of cases, the test won’t adequately cover those behaviors. 
 *
 */

/**
 * Creates a mock node that returns a focus function that does nothing. Required to avoid errors on input.focus
 * 
 * react-test-renderer doesn’t actually use the browser (or jsdom), it does not return real elements/tags
 * If your components crash expecting some methods (like focus()), a workaround was implemented  * in #7649 and will
 * be part of 15.4.0. It will work like this: https://github.com/facebook/react/issues/7740
 * Anything can be returned from this function.
 * @return {Function} 
 */
function createNodeMock() {
  return {
    focus() {
      // Do nothing
    }
  }
}

/******************************************************************
* App Snapshot Tests 
******************************************************************/

test('App renders correctly', () => {
  	const tree = renderer.create(
    	<App />, {createNodeMock}
  	).toJSON();
  	expect(tree).toMatchSnapshot();
});

/******************************************************************
* SearchSelectContainer Snapshot Tests 
******************************************************************/

// I guess this is redundent since App holds same snapshot? Still learning.....
test('SearchSelectContainer renders correctly', () => {
  	const tree = renderer.create(
    	<SearchSelectContainer />, {createNodeMock}
  	).toJSON();
  	expect(tree).toMatchSnapshot();
});