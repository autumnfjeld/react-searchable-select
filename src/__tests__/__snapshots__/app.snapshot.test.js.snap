// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="app-container"
>
  <header
    className="app-header"
  >
    <h1
      className="app-title"
    >
      MARVEL
    </h1>
    <div>
      <div
        className="search-select-container"
      >
        <form>
          <input
            autoFocus={true}
            className="character-search-input"
            onChange={[Function]}
            placeholder="Search.."
            type="text"
            value=""
          />
          <div
            className=""
          />
        </form>
      </div>
    </div>
  </header>
  <div
    className="app-content"
  />
</div>
`;

exports[`SearchSelectContainer renders correctly 1`] = `
<div
  className="search-select-container"
>
  <form>
    <input
      autoFocus={true}
      className="character-search-input"
      onChange={[Function]}
      placeholder="Search.."
      type="text"
      value=""
    />
    <div
      className=""
    />
  </form>
</div>
`;
